Tasks/Maven/maven2.js

```diff
@@ -48,40 +50,161 @@ var publishJUnitResults = tl.getInput('publishJUnitResults');
+        sqArguments = createSonarQubeArgs(sqEndpoint.Url, sqEndpoint.Username, sqEndpoint.Password);
+    }
+
+    sqArguments = mavenOptions + " " + sqArguments;
+    tl.debug("Running Maven with goal sonar:sonar and options " + sqArguments);
+
+    var mvnsq = new tl.ToolRunner(mvntool);
+    mvnsq.arg('-f');
+    mvnsq.arg(mavenPOMFile);
+    mvnsq.arg(sqArguments);
+    mvnsq.arg("sonar:sonar");
+
+    return mvnsq;
+}
+
+function getEndpointDetails(inputFieldName) {
+    var errorMessage = "Could not decode the generic endpoint. Please ensure you are running the latest agent (min version 0.3.0)"
+    if (!tl.getEndpointUrl) {
+        throw new Error(errorMessage);
+    }
+
+    var genericEndpoint = tl.getInput(inputFieldName);
+    if (!genericEndpoint) {
+        throw new Error(errorMessage);
     }
-    else {
-      tl.debug('No pattern found in testResultsFiles parameter');
-      var matchingTestResultsFiles = [testResultsFiles];
+
+    hostUrl = tl.getEndpointUrl(genericEndpoint, false);
+    var auth = tl.getEndpointAuthorization(genericEndpoint, false);
+
+    if (auth.scheme != "UsernamePassword") {
+        throw new Error("The authorization scheme " + auth.scheme + " is not supported for a SonarQube endpoint. Please use a username and a password.");
     }
 
-    if(!matchingTestResultsFiles) {
-      tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');  
-      return 0;
+    hostUsername = auth.parameters.Username;
+    hostPassword = auth.parameters.Password;
+
+    return {
+        "Url": hostUrl,
+        "Username": hostUsername,
+        "Password": hostPassword
+    };
+}
+
+function createSonarQubeArgs(sqHostUrl, sqHostUsername, sqHostPassword, sqDbUrl, sqDbUsername, sqDbPassword) {
+    var sqArgs = '-Dsonar.host.url="' + sqHostUrl + '" ';
+    if (sqHostUsername) {
+        sqArgs += '-Dsonar.login="' + sqHostUsername + '" ';
+    }
+    if (sqHostPassword) {
+        sqArgs += '-Dsonar.password="' + sqHostPassword + '" ';
+    }
+    if (sqDbUrl) {
+        sqArgs += '-Dsonar.jdbc.url="' + sqDbUrl + '" ';
+    }
+    if (sqDbUsername) {
+        sqArgs += '-Dsonar.jdbc.username="' + sqDbUsername + '" ';
+    }
+    if (sqDbPassword) {
+        sqArgs += '-Dsonar.jdbc.password="' + sqDbPassword + '" ';
     }
 
-    var tp = new tl.TestPublisher("JUnit");
-    tp.publish(matchingTestResultsFiles, false, "", "");
-  } 
+    return sqArgs;
 }
 
+/*
+Maven task orchestration:
+
+1. Check that maven exists 
+2. Run maven with the user goals. Compilation or test errors will cause this to fail
+3. Always try to publish tests results 
+4. Always try to run the SonarQube analysis if it is enabled. In case the build has failed, the analysis 
+will still succeed but the report will have less data. 
+5. If #2 above failed, exit with an error code to mark the entire step as failed. Same for #4.
+
+*/
+
+
+var runFailed = false;
+
 mvnv.exec()
```

Comment:
As Duncan pointed out, this call is redundant because the Maven task requires "Maven" via the capabilities model. I left it because the other Java build tasks (ant and gradle) follow the same pattern, although the code would be cleaner without it.